pkg/kubelet/kubelet.go

```diff
@@ -699,11 +699,19 @@ func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
 		}
 		// TODO(roberthbailey): Can we do this without having credentials to talk
 		// to the cloud provider?
-		instanceID, err := instances.ExternalID(kl.hostname)
+		externalID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get external ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = externalID
+		// TODO: We can't assume that the node has credentials to talk to the
+		// cloudprovider from arbitrary nodes. At most, we should talk to a
+		// local metadata server here.
+		instanceID, err := instances.InstanceID(node.Name)
 		if err != nil {
 			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
 		}
-		node.Spec.ExternalID = instanceID
+		node.Spec.ProviderID = kl.cloud.ProviderName() + "://" + instanceID
 	} else {
 		node.Spec.ExternalID = kl.hostname
```

Comment:
Do we not want/need to set the ProviderID when no cloud provider is specified? If ExternalID is deprecated and will be removed, then we are going to lose this "else" case once it is removed.